server/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -1416,7 +1418,57 @@ public void openEngineAndSkipTranslogRecovery() throws IOException {
         getEngine().skipTranslogRecovery();
     }
 
-    private void innerOpenEngineAndTranslog() throws IOException {
+    /**
+     * Recovers a replica locally up to the global checkpoint before starting a peer recovery.
+     */
+    public void prepareShardForPeerRecovery() {
+        if (state != IndexShardState.RECOVERING) {
+            throw new IndexShardNotRecoveringException(shardId, state);
+        }
+        if (recoveryState.getRecoverySource() instanceof RecoverySource.PeerRecoverySource == false) {
+            assert false : "only peer recovery needs to recover locally up to the global checkpoint";
+            throw new IllegalStateException("only peer recovery needs to recover locally up to the global checkpoint");
+        }
+        try {
+            final long globalCheckpoint;
+            final Engine engine;
+            synchronized (mutex) {
+                String translogUUID = store.readLastCommittedSegmentsInfo().getUserData().get(Translog.TRANSLOG_UUID_KEY);
+                globalCheckpoint = Translog.readGlobalCheckpoint(translogConfig.getTranslogPath(), translogUUID);
+                List<IndexCommit> commits = DirectoryReader.listCommits(store.directory());
+                IndexCommit safeCommit = CombinedDeletionPolicy.findSafeCommitPoint(commits, globalCheckpoint);
+                SequenceNumbers.CommitInfo commitInfo = SequenceNumbers.loadSeqNoInfoFromLuceneCommit(safeCommit.getUserData().entrySet());
+                boolean shouldRecoverLocally = commitInfo.maxSeqNo <= globalCheckpoint && commitInfo.localCheckpoint < globalCheckpoint;
+                if (shouldRecoverLocally && indexSettings.getIndexMetaData().getState() == IndexMetaData.State.OPEN) {
+                    engine = innerOpenEngineAndTranslog();
+                } else {
+                    engine = null;
+                }
+            }
+            if (engine != null) {
+                try {
+                    indexShardOperationPermits.blockOperations(-1, TimeUnit.SECONDS, () -> {
+                        final Engine.TranslogRecoveryRunner translogRecoveryRunner =
+                            (e, snapshot) -> runTranslogRecovery(e, snapshot, Engine.Operation.Origin.LOCAL_RESET, () -> { });
```

Comment:
why use this instead of `LOCAL_TRANSLOG_RECOVERY` as origin? I'm not sure which of the two is more appropriate, but was wondering about what motivated your choice here